### PR TITLE
o/daemon: validation sets api and basic spread test

### PR DESCRIPTION
**Pull request by stolowski**
This PR implements /v2/validation-sets endpoint with GET method for querying and POST for applying (updating/removing validation) sets. The code updates and reads validation-set-tracking map in the state (via assertstate methods from #9391), but there a TODOs for actual logic related to tracked validation sets.

The spread test is a basic check of all commands, but will very likely need to evolve once wa start using store and real validations sets.

### CODE REVIEW

File: daemon/api.go [hunk 1]
```diff
@@ -154,6 +156,17 @@ var (
 		GET:    getAliases,
 		POST:   changeAliases,
 	}
+
+	validationSetsListCmd = &Command{
```
Comment (pedronis) [inline daemon/api.go:160]: we tend to put these nowadays in the matching api_*go file, not here

File: daemon/api_validate.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+		key = fmt.Sprintf("%s=%d", key, sequence)
+	}
+	return key
+}
+
+func validationSetNotFound(accountID, name string, sequence int) Response {
+	res := &errorResult{
+		Message: "validation set not found",
+		Kind:    client.ErrorKindValidationSetNotFound,
+		Value:   validationSetKeyString(accountID, name, sequence),
```
Comment (pedronis) [inline daemon/api_validate.go:67]: we need to think a bit about this, should it be a string? or some kind of maps with exploded key components?

File: daemon/api_validate.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	for i, vs := range names {
+		tr := validationSets[vs]
+		// TODO: evaluate against installed snaps
+		var valid bool
+		modeStr, err := modeString(tr.Mode)
+		if err != nil {
+			return InternalError(err.Error())
+		}
+		results[i] = validationSetResult{
+			ValidationSet: validationSetKeyString(tr.AccountID, tr.Name, tr.PinnedAt),
```
Comment (pedronis) [inline daemon/api_validate.go:118]: I missed that we would incorporate pinned-at in the string itself, I wonder if we should make it a separate field instead

File: daemon/api_validate.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var tr assertstate.ValidationSetTracking
+	err := assertstate.GetValidationSet(st, accountID, name, &tr)
+	if err == state.ErrNoState || (err == nil && sequence != 0 && sequence != tr.PinnedAt) {
+		return validationSetNotFound(accountID, name, sequence)
```
Comment (pedronis) [inline daemon/api_validate.go:163]: if we don't find it locally we need to try to look for it in the store, that can be a TODO for now

File: daemon/api_validate.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	switch req.Mode {
+	case "monitor":
+		mode = assertstate.Monitor
+	case "enforce":
+		mode = assertstate.Enforce
+	default:
+		return BadRequest("invalid mode %q", req.Mode)
+	}
+
+	// TODO: check that it exists in the store?
```
Comment (pedronis) [inline daemon/api_validate.go:235]: yes here we need to see if we have the needed assertion locally already if pinned, and then also check with the store if there is something newer there, or what is the latest there if the assertion is not pinned

File: daemon/api_validate.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+	tr := assertstate.ValidationSetTracking{
+		AccountID: accountID,
+		Name:      name,
+		Mode:      mode,
+		// note, Sequence may be 0, meaning not pinned.
+		PinnedAt: req.Sequence,
+	}
```
Comment (pedronis) [inline daemon/api_validate.go:243]: if the mode is enforced here we need to check that the assertion is valid before saving it, for one that comes from the store we would also add it to the assertion db,  we might need more helpers in assertstate for this

File: daemon/api_validate.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+	assertstate.UpdateValidationSet(st, &tr)
+	return SyncResponse(nil, nil)
+}
+
+func forgetValidationSet(st *state.State, accountID, name string, sequence int) Response {
+	// check if it exists first
```
Comment (pedronis) [inline daemon/api_validate.go:249]: maybe a comment here that the state is already locked

**Review by pedronis — COMMENTED** [review]
did a pass, some comments,  some are really TODOs for  later follow ups

File: daemon/api_validate.go (hunk 3 above)
Comment (stolowski) [reply daemon/api_validate.go:118]: Yes I think separate fields also for account, name and pinned-at make sense.

File: daemon/api_validate.go (hunk 2 above)
Comment (stolowski) [reply daemon/api_validate.go:67]: I agree, I think it makes sense to make life easier for the clients.

File: client/validate.go [hunk 8]
```diff
@@ -36,10 +36,12 @@ type ValidateApplyOptions struct {
 
 // ValidationSetResult holds information about a single validation set.
 type ValidationSetResult struct {
-	ValidationSet string `json:"validation-set,omitempty"`
-	Mode          string `json:"mode"`
-	Sequence      int    `json:"sequence,omitempty"`
-	Valid         bool   `json:"valid"`
+	AccountID string `json:"account-id"`
```
Comment (pedronis) [inline client/validate.go:39]: maybe a comment before this like

// set sequence key and optional pinned-at (=)

File: client/validate.go [hunk 9]
```diff
@@ -36,10 +36,12 @@ type ValidateApplyOptions struct {
...
+	Name      string `json:"name"`
+	PinnedAt  int    `json:"pinned-at,omitempty"`
+	Mode      string `json:"mode"`
```
Comment (pedronis) [inline client/validate.go:42]: and here:

// set current state

File: daemon/api_validate.go [hunk 10]
```diff
@@ -74,10 +74,17 @@ func validationSetKeyString(accountID, name string, sequence int) string {
 }
 
 func validationSetNotFound(accountID, name string, sequence int) Response {
+	v := map[string]interface{}{
+		"account": accountID,
```
Comment (pedronis) [inline daemon/api_validate.go:72]: s/account/account-id/

**Review by pedronis — APPROVED** [review]
looks good, small change needed in the error value and some other minor comments

File: daemon/api_validate.go (hunk 10 above)
Comment (stolowski) [reply daemon/api_validate.go:72]: Thanks, missed that one.

File: daemon/api_validate_test.go [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+
+	"github.com/snapcore/snapd/daemon"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+var _ = check.Suite(&apiValidationSetsSuite{})
+
+type apiValidationSetsSuite struct {
+	daemon.APIBaseSuite
```
Comment (pedronis) [inline daemon/api_validate_test.go:38]: this now can be apiBaseSuite, only change is that some method on it are unexported (again)

File: daemon/api_validate_test.go [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+
+	d *daemon.Daemon
+}
+
+func (s *apiValidationSetsSuite) SetUpTest(c *check.C) {
+	s.APIBaseSuite.SetUpTest(c)
+	s.d = s.DaemonWithStore(c, s)
```
Comment (pedronis) [inline daemon/api_validate_test.go:45]: also this is probably not needed anymore

File: daemon/api_validate.go [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+	if req.Sequence < 0 {
+		return BadRequest("invalid sequence argument: %d", req.Sequence)
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if req.Action == "forget" {
+		return forgetValidationSet(st, accountID, name, req.Sequence)
```
Comment (mvo5) [inline daemon/api_validate.go:238]: (nitpick) I personally would probably split the rest of the function into a helper similar to forgetValidationSet and make the if req.Action() a switch/case. But this here is also fine.

File: daemon/api_validate_test.go [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+func (s *apiValidationSetsSuite) SetUpTest(c *check.C) {
+	s.apiBaseSuite.SetUpTest(c)
+	d := s.daemon(c)
+	d.Overlord().Loop()
+	defer d.Overlord().Stop()
+}
+
+func (s *apiValidationSetsSuite) TearDownTest(c *check.C) {
```
Comment (mvo5) [inline daemon/api_validate_test.go:48]: This should not be needed if just the base is called or am I missing something here?

File: daemon/api_validate_test.go [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+
+type apiValidationSetsSuite struct {
+	apiBaseSuite
+}
+
+func (s *apiValidationSetsSuite) SetUpTest(c *check.C) {
+	s.apiBaseSuite.SetUpTest(c)
+	d := s.daemon(c)
+	d.Overlord().Loop()
+	defer d.Overlord().Stop()
```
Comment (mvo5) [inline daemon/api_validate_test.go:45]: Should this be `s.AddCleanup(d.Overlord().Stop)` ? Using a defer here that stops overlord as soon as the function exits? Or am I missing something? If so, maybe a short comment?

File: daemon/api_validate_test.go [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+			body = fmt.Sprintf(`{"action":"apply","mode":"%s", "sequence":%s}`, tc.mode, tc.sequence)
+		} else {
+			body = fmt.Sprintf(`{"action":"apply","mode":"%s"}`, tc.mode)
+		}
+		req, err := http.NewRequest("POST", fmt.Sprintf("/v2/validation-sets/%s", tc.validationSet), strings.NewReader(body))
+		c.Assert(err, check.IsNil)
+		rsp := s.req(c, req, nil).(*daemon.Resp)
+
+		c.Check(rsp.Type, check.Equals, daemon.ResponseTypeError, check.Commentf("case #%d", i))
+		c.Check(rsp.Status, check.Equals, tc.status, check.Commentf("case #%d", i))
```
Comment (mvo5) [inline daemon/api_validate_test.go:423]: Tests look nice!

**Review by mvo5 — COMMENTED** [review]
Looks very good, one questions in the test code about the defer that feels important, once that is resolved/answer this can be merged I think

File: daemon/api_validate_test.go (hunk 14 above)
Comment (stolowski) [reply daemon/api_validate_test.go:48]: You are right, thanks!

File: daemon/api_validate.go (hunk 13 above)
Comment (stolowski) [reply daemon/api_validate.go:239]: Done

**Review by mvo5 — APPROVED** [review]
Thank you!

**Pull request merged**